site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -1384,6 +1388,45 @@ private function getStats($gradeable, &$sections, $graders=array(), $total_users
             }
         }
     }
+
+    /**
+     * Check if the current user is allowed to grade the given gradeable/user pair
+     * @param string $gradeable_id
+     * @param string $user_id
+     * @return bool
+     */
+    public function canIGradeThis(string $gradeable_id, string $user_id): bool {
+        $gradeable = $this->core->getQueries()->getGradeable($gradeable_id, $user_id);
+        if ($gradeable === null) {
+            return false;
+        }
+
+        //checks if user has permission
+        if ($this->core->getUser()->getGroup() === 4) {
+            if(!$gradeable->getPeerGrading()) {
+                return false;
+            } else {
+                $user_ids_to_grade = $this->core->getQueries()->getPeerAssignment($gradeable->getId(), $this->core->getUser()->getId());
+                if(!in_array($user_id, $user_ids_to_grade)) {
+                    return false;
+                }
+            }
+        } else if ($this->core->getUser()->getGroup() === 3) {
+            if ($gradeable->isGradeByRegistration()) {
+                $sections = $this->core->getUser()->getGradingRegistrationSections();
+                $users_to_grade = $this->core->getQueries()->getUsersByRegistrationSections($sections);
+            } else {
+                $sections = $this->core->getQueries()->getRotatingSectionsForGradeableAndUser($gradeable_id, $this->core->getUser()->getId());
+                $users_to_grade = $this->core->getQueries()->getUsersByRotatingSections($sections);
+            }
+            $user_ids_to_grade = array_map(function(User $user) { return $user->getId(); }, $users_to_grade);
+            if (!in_array($user_id, $user_ids_to_grade)) {
+                return false;
+            }
```

Comment:
Just foreach through the array, as using array_map and then in_array has an O(n<sup>2</sup>) runtime if false and we can reduce it to just O(n).

```
foreach ($users_to_grade as $user) {
    if ($user->getId() === $user_id) {
        return true;
    }
}
return false;
```